plugins/MumbleAPI_v_1_0_x.h

```diff
@@ -0,0 +1,457 @@
+	/// 	allocated if this function returns STATUS_OK.
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer may
+	/// 	be accessed.
+	mumble_error_t (PLUGIN_CALLING_CONVENTION *getMumbleSetting_string)(mumble_plugin_id_t callerID, mumble_settings_key_t key, const char **outValue);
+
+
+	/// Sets the value of the setting with the given key. Note that this function can only be used for settings whose value
+	/// is a bool!
+	///
+	/// @param callerID The ID of the plugin calling this function
+	/// @param key The key to the desired setting
+	/// @param value The value that should be set for the given setting
+	/// @returns The error code. If everything went well, STATUS_OK will be returned.
+	mumble_error_t (PLUGIN_CALLING_CONVENTION *setMumbleSetting_bool)(mumble_plugin_id_t callerID, mumble_settings_key_t key, bool value);
+
+	/// Sets the value of the setting with the given key. Note that this function can only be used for settings whose value
+	/// is an int!
+	///
+	/// @param callerID The ID of the plugin calling this function
+	/// @param key The key to the desired setting
+	/// @param value The value that should be set for the given setting
+	/// @returns The error code. If everything went well, STATUS_OK will be returned.
+	mumble_error_t (PLUGIN_CALLING_CONVENTION *setMumbleSetting_int)(mumble_plugin_id_t callerID, mumble_settings_key_t key, int value);
+
+	/// Sets the value of the setting with the given key. Note that this function can only be used for settings whose value
+	/// is a double!
+	///
+	/// @param callerID The ID of the plugin calling this function
+	/// @param key The key to the desired setting
+	/// @param value The value that should be set for the given setting
+	/// @returns The error code. If everything went well, STATUS_OK will be returned.
+	mumble_error_t (PLUGIN_CALLING_CONVENTION *setMumbleSetting_double)(mumble_plugin_id_t callerID, mumble_settings_key_t key, double value);
+
+	/// Sets the value of the setting with the given key. Note that this function can only be used for settings whose value
+	/// is a string!
+	///
+	/// @param callerID The ID of the plugin calling this function
+	/// @param key The key to the desired setting
+	/// @param value The value that should be set for the given setting
+	/// @returns The error code. If everything went well, STATUS_OK will be returned.
+	mumble_error_t (PLUGIN_CALLING_CONVENTION *setMumbleSetting_string)(mumble_plugin_id_t callerID, mumble_settings_key_t key, const char *value);
+
+
+
+
+	// -------- Miscellaneous --------
+	
+	/// Sends the provided data to the provided client(s). This kind of data can only be received by another plugin active
+	/// on that client. The sent data can be seen by any active plugin on the receiving client. Therefore the sent data
+	/// must not contain sensitive information or anything else that shouldn't be known by others.
+	///
```

Comment:
Add a note about ratelimit.
also, is there a way for a plugin to retrieve the servers ratelimit ?
(btw, i am going to abuse this part of the api lol)

Reply:
Good point :+1: 

> also, is there a way for a plugin to retrieve the servers ratelimit ?

I did not add such mechanism directly. _Maybe_ it is possible to query the values via Ice. But I assume you were thinking about a way through the plugin API. That is not supported.
If it is useful though, we can talk about adding such functionality once the plugin framework landed in master :)

> btw, i am going to abuse this part of the api lol

How are you going to _abuse_ this? ^^

Reply:
> If it is useful though, we can talk about adding such functionality once the plugin framework landed in master :)

I would advise against exposing server sided internal settings for security reasons.
Maybe an option without exposing it could be to return a error message to the plugin (the API call is already blocking, correct?). This would also allow the plugin to try to resend messages that are expected to be transmitted reliably (if needed), or do other nice things.

Reply:
> once the plugin framework landed in master

hope this happens soon :D

> How are you going to abuse this? ^^

sending and receiving arbitrary data between clients? i always hoped this would land in some form via a plugin system.
i am thinking of prototyping simple file-transfers, video steams etc... and possible with "just" putting a dll into a folder ^^

Reply:
> hope this happens soon :D

Me too :D
But I am optimistic. I just finished my review yesterday. Next up I'll have to fix my review comments and then it can be merged :+1:

Reply:
ok, so, I thought about it some more. We need to define a max size for the data field, bc not limiting would mean each message can be some ~15MB. If I understand the code right (https://github.com/Krzmbrzl/mumble/blob/plugin-API/src/Connection.cpp#L169) . The rate-limit of 4 per second would mean 60MB/s.
If I am not wrong.

Reply:
Hm yeah that is probably correct. I did not think about that when I implemented this :thinking: 

What would you consider to be a good size limit for such a message? Something like ~5KB?

@hbeni afaik your plugin makes use of this feature already. Do you have an estimate how large the largest packet of yours ends up being?

Reply:
Yes, i make extensive use of the plugin messages.
I would guess, average packet size is about 300 bytes, with possible occasional spikes up, but in my usecase i should consistently stay well below 1k. It depends on user input, however :)

Having a server side configurable limit for this may be a useful thing.

Reply:
Ah and thinking about it: the data-id field of that package may be subject to limits too - currently i think a plugin can send arbitary ids too.

Reply:
> Ah and thinking about it: the data-id field of that package may be subject to limits too - currently i think a plugin can send arbitary ids too.

Yes indeed :sweat_smile: 

> Having a server side configurable limit for this may be a useful thing.

I think that makes it more complicated than it needs to be. If a plugin author knows that this message may at most be x bytes long, they can engineer their plugin to either stay below that limit or to split their messages in parts and send these individually. If the limit is variable that would make something like this more complicated :point_up:

Reply:
can't i kill a server with a malicious plugin then, easily?

Reply:
Ah, you talk about a hard coded limit. That would be OK i think.
Should be mentioned in the header file :)

Reply:
> Something like ~5KB?

> should consistently stay well below 1k

As a rule of thumb, a network packet should be at most be 1k. Making this configurable would be nice too, but we already have ratelimit, so i would hardcode it.

What are the limits for the id? I would not go above 50 chars for that one.

Reply:
Okay then 1K it is :+1: 

> What are the limits for the id? I would not go above 50 chars for that one.

Yeah I suppose 50 characters should be sufficient. But then again we are only trying to prevent abusing this field, so I would suggest staying on the safe side of things and setting the limit to 100 bytes for the ID field :thinking:

Reply:
For the ID: we are talking about the mumble plugin ID, not the DataID field?
`mumble_error_t (PLUGIN_CALLING_CONVENTION *sendData)(mumble_plugin_id_t callerID, mumble_connection_t connection, const mumble_userid_t *users, size_t userCount, const uint8_t *data, size_t dataLength, const char *dataID);`

I use the `dataID` characters to encode some data, so this should allow more space than 100 bytes. 25 chars should be fine for me, but i would advice for more, say 50 characters.

Reply:
nevermind. I need more coffe or a break :D
I confused bits with bytes. 100 bytes is of course sufficient.
Sorry for the noise.

Reply:
> I use the dataID characters to encode some dynamic data

Uhm this is not how this is intended to be used. This really should only be an ID and not part of the transferred data. So you should only use an ID that makes it clear that this message belongs to your plugin and that your plugin furthermore is able to figure out what kind of data it receives. Everything else should be in the data itself :point_up:

Reply:
Yes, agreed.
I encode for example identifiers/selectors like this: `FGCOM:UPD_COM:<n>` to update the radio data of Radio `<n>`.

Reply:
I suggest sending `FGCOM:UDP_COM:CHANGE_RADIO` as the ID and the frequency as (part of) the data instead :point_up:

Reply:
The point is that the packet contains all the data for that specific radio. It’s not just the  frequency, and there can be lots of radios be registered.

Reply:
This could still be encoded in the sent data :shrug:

Reply:
It could, yes of course.
I found it just more convenient to do it this way.

is there serious harm to be expected if i leave it as is?

Reply:
> is there serious harm to be expected if i leave it as is?

The only harm I see in this is this: The more of these IDs plugins use, the higher the risk of name clashes between plugins which could lead to errors since now a plugin might attempt to handle data that was not meant for it.
Given that you are using a prefix indicating your plugin's name though (as every plugin should do), I expect the problem to not really surface.

So _effectively_ I don't see a big potential for any harm that this could cause :shrug:

Reply:
Very good!

I think it would be a good idea to document it somewhere that the dataID should start with an unique prefix.
A good place is in the header file, but also the example plugin.